odoo/cli/db.py

```diff
@@ -0,0 +1,152 @@
+                 "(without extension)"
+        )
+        load.add_argument('dump_file', help="zip or pg_dump file to load")
+
+        dump = subs.add_parser(
+            "dump", help="Create a dump with filestore.",
+            description="Creates a dump file. The dump is always in zip format "
+                        "(with filestore), to get a no-filestore format use "
+                        "pg_dump directly.")
+        dump.set_defaults(func=self.dump)
+        dump.add_argument('database', help="database to dump")
+        dump.add_argument(
+            'dump_path', nargs='?', default='-',
+            help="if provided, database is dumped to specified path, otherwise "
+                 "or if `-`, dumped to stdout",
+        )
+
+        duplicate = subs.add_parser("duplicate", help="Duplicate a database including filestore.")
+        duplicate.set_defaults(func=self.duplicate)
+        duplicate.add_argument(
+            '-f', '--force', action='store_const', default=False, const=True,
+            help="delete `target` database before copying if it exists"
+        )
+        duplicate.add_argument("source")
+        duplicate.add_argument("target", help="database to copy `source` to, must not exist unless `-f` is specified in which case it will be dropped first")
+
+        rename = subs.add_parser("rename", help="Rename a database including filestore.")
+        rename.set_defaults(func=self.rename)
+        rename.add_argument(
+            '-f', '--force', action='store_const', default=False, const=True,
+            help="delete `target` database before renaming if it exists"
+        )
+        rename.add_argument('source')
+        rename.add_argument("target", help="database to rename `source` to, must not exist unless `-f` is specified, in which case it will be dropped first")
+
+        drop = subs.add_parser("drop", help="Delete a database including filestore")
+        drop.set_defaults(func=self.drop)
+        drop.add_argument("database", help="database to delete")
+
+        args = parser.parse_args(cmdargs)
+
+        config.parse_config([
+            val
+            for k, v in vars(args).items()
+            if v is not None
+            if k in ['config', 'data_dir', 'addons_path'] or k.startswith(('db_', 'pg_'))
+            for val in [
+                '--data-dir' if k == 'data_dir'\
+                    else '--addons-path' if k == 'addons_path'\
+                    else f'--{k}',
+                v,
+            ]
+        ])
+        # force db management active to bypass check when only a
+        # `check_db_management_enabled` version is available.
+        config['list_db'] = True
+        report_configuration()
+
+        args.func(args)
+
+    def load(self, args):
+        url = urllib.parse.urlparse(args.dump_file)
+        if url.scheme:
+            stderr.write(f"Fetching {args.dump_file}...")
+            r = requests.get(args.dump_file, timeout=10, stream=True)
```

Comment:
I don't really understand how it would work testing it i have 
```
TypeError: expected str, bytes or os.PathLike object, not HTTPResponse
```

Reply:
Hmmm guessing the original put everything in memory wrapped in a `BytesIO`, but then I figured the response was file-like so I could just pass it directly to `restore_db`, and then I forgot to test that, and it doesn't work because the http stream is not seekable.

Reply:
http stream should be seekable but it is passed directly to zipfile.zipfile and it's expecting a filename.

Reply:
> http stream should be seekable

It's not, I checked. Even passing it directly to `ZipFile` blows up,

>  but it is passed directly to zipfile.is_zipfile and it's expecting a filename.

Nope, takes a file-like object or a filename, but apparently if it gets a non-seekable file-like object it just returns `False` rather than an error, so we get the second path in `restore_db`.

But that means I should probably `is_zipfile` before calling `restore_db` otherwise it treats whatever nonsense we got as a raw postgres dump file and tries to `pg_restore` it. Which is not really necessary on the command line because you can just `pg_restore` it yourself like a big boy.

> My tests command line: `./odoo-bin db load http://runbot57.odoo.com/runbot/static/build/18642143-master/logs/18642143-master-all.zip`

Yeah that's what I checked with (and pretty much the entire point of the feature). I reversed it.